contracts/chain-adapters/ZkStack_Adapter.sol

```diff
@@ -0,0 +1,269 @@
+
+    // We need to pay a base fee to the operator to include our L1 --> L2 transaction.
+    // https://era.zksync.io/docs/dev/developer-guides/bridging/l1-l2.html#getting-the-base-cost
+
+    // Limit on L2 gas to spend.
+    uint256 public immutable L2_GAS_LIMIT; // typically 2_000_000
+
+    // How much gas is required to publish a byte of data from L1 to L2. 800 is the required value
+    // as set here https://github.com/matter-labs/era-contracts/blob/6391c0d7bf6184d7f6718060e3991ba6f0efe4a7/ethereum/contracts/zksync/facets/Mailbox.sol#L226
+    // Note, this value can change and will require an updated adapter.
+    uint256 public immutable L1_GAS_TO_L2_GAS_PER_PUB_DATA_LIMIT; // Typically 800
+
+    // This address receives any remaining fee after an L1 to L2 transaction completes.
+    // If refund recipient = address(0) then L2 msg.sender is used, unless msg.sender is a contract then its address
+    // gets aliased.
+    address public immutable L2_REFUND_ADDRESS;
+
+    // L2 chain id
+    uint256 public immutable CHAIN_ID;
+
+    // BridgeHub address
+    BridgeHubInterface public immutable BRIDGE_HUB;
+
+    // Set l1Weth at construction time to make testing easier.
+    WETH9Interface public immutable L1_WETH;
+
+    // SharedBridge address, which is read from the BridgeHub at construction.
+    address public immutable SHARED_BRIDGE;
+
+    event ZkStackMessageRelayed(bytes32 canonicalTxHash);
+    error ETHGasTokenRequired();
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l1Weth WETH address on L1.
+     * @param _l2RefundAddress address that recieves excess gas refunds on L2.
+     */
+    constructor(
+        uint256 _chainId,
+        BridgeHubInterface _bridgeHub,
+        WETH9Interface _l1Weth,
+        address _l2RefundAddress,
+        uint256 _l2GasLimit,
+        uint256 _l1GasToL2GasPerPubDataLimit
+    ) {
+        CHAIN_ID = _chainId;
+        BRIDGE_HUB = _bridgeHub;
+        L1_WETH = _l1Weth;
+        L2_REFUND_ADDRESS = _l2RefundAddress;
+        L2_GAS_LIMIT = _l2GasLimit;
+        L1_GAS_TO_L2_GAS_PER_PUB_DATA_LIMIT = _l1GasToL2GasPerPubDataLimit;
+        SHARED_BRIDGE = BRIDGE_HUB.sharedBridge();
+        address gasToken = BRIDGE_HUB.baseToken(CHAIN_ID);
+        if (gasToken != address(1)) {
+            revert ETHGasTokenRequired();
+        }
+    }
+
+    /**
+     * @notice Send cross-chain message to target on ZkStack.
+     * @dev The HubPool must hold enough ETH to pay for the L2 txn.
+     * @param target Contract on L2 that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes memory message) external payable override {
+        uint256 txBaseCost = _computeETHTxCost(L2_GAS_LIMIT);
+
+        // Returns the hash of the requested L2 transaction. This hash can be used to follow the transaction status.
+        bytes32 canonicalTxHash = BRIDGE_HUB.requestL2TransactionDirect{ value: txBaseCost }(
```

Comment:
You note in the interface for `requestL2TransactionDirect` that:

> This means this is not ideal for contract calls, as the contract would have to handle token allowance of the base Token.

And yet you're making a contract call here, right?

Reply:
This natspec is copied directly from their docs. This was also confusing to me, but I copied it just for posterity.

I used [this example](https://etherscan.io/address/0x3766Eb5F07DBc60d39a2059A9A29fD9b7D3C356D#code) as a guide to how these functions are used for both token and native asset bridging.